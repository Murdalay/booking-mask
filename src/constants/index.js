import keyMirror from 'fbjs/lib/keyMirror';

/**
 * @namespace Constants
 * @desc App constants
 */

/**
 * @constant {Object} ActionTypes
 * @memberof Constants
 */
export const ActionTypes = keyMirror({
  EXCEPTION: undefined,
  TOGGLE_ROUND_TRIP: undefined,
  BOOKING_TOGGLE_PASSANGERS_DROPDOWN: undefined,
  BOOKING_CHANGE_PASSANGERS_NUMBER: undefined,
  BOOKING_CHANGE_FROM: undefined,
  BOOKING_CHANGE_TO: undefined,
  BOOKING_CHANGE_FLIGHT_DATES: undefined,
  BOOKING_SUBMIT: undefined,
  BOOKING_SUBMIT_SUCCESS: undefined,
  BOOKING_SET_INVALID: undefined,
  BOOKING_UNSET_INVALID: undefined,
  BOOKING_URL_READY: undefined,
  AIROPORTS_LIST_GET: undefined,
  AIROPORTS_LIST_GET_SUCCESS: undefined,
  AIROPORTS_LIST_GET_FAILURE: undefined,
  SHOW_ALERT: undefined,
  HIDE_ALERT: undefined,
});

/**
 * @constant {Object} STATUS
 * @memberof Constants
 */
export const STATUS = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
