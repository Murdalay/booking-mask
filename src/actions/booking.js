// @flow
/**
 * @module Actions/Booking
 * @desc Booking Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const {
  bookingChangePassengersNumber: changePassengersNumber,
  bookingChangeFrom: changeFrom,
  bookingChangeTo: changeTo,
  bookingChangeFlightDates: changeFlightDates,
  bookingSubmit: submit,
  bookingUnsetInvalid: unsetInvalid,
  bookingTogglePassengersDropdown: togglePassengersDropdown,
} = createActions({
  [ActionTypes.BOOKING_GET_REPOS]: (query: string) => ({ query }),
  [ActionTypes.BOOKING_CHANGE_PASSENGERS_NUMBER]: (data: {
    adults: number,
    children: number,
    infants: number,
  }) => ({ data }),
  [ActionTypes.BOOKING_CHANGE_FROM]: data => ({ code: data }),
  [ActionTypes.BOOKING_CHANGE_TO]: data => ({ code: data }),
  [ActionTypes.BOOKING_CHANGE_FLIGHT_DATES]: (data: { startDate: string, endDate: string }) => data,
  [ActionTypes.BOOKING_SUBMIT]: () => ({}),
  [ActionTypes.BOOKING_UNSET_INVALID]: () => ({}),
  [ActionTypes.BOOKING_TOGGLE_PASSENGERS_DROPDOWN]: () => ({}),
});
