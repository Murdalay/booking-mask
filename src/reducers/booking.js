import { handleActions } from 'redux-actions';
import immutable from 'immutability-helper';
import { ActionTypes } from 'constants/index';

export const bookingState = {
  passengers: {
    total: 1,
    adults: 1,
    children: 0,
    infants: 0,
  },
  flightDates: {
    departure: null,
    return: null,
  },
  passengersDropdownIsOpen: false,
  isRoundTrip: false,
  from: null,
  to: null,
  invalid: null,
  url: null,
};

export default {
  booking: handleActions(
    {
      [ActionTypes.BOOKING_CHANGE_FROM]: (state, { payload }) => {
        return immutable(state, {
          from: { $set: payload.code || null },
          url: { $set: null },
        });
      },
      [ActionTypes.BOOKING_CHANGE_TO]: (state, { payload }) => {
        return immutable(state, {
          to: { $set: payload.code || null },
          url: { $set: null },
        });
      },
      [ActionTypes.TOGGLE_ROUND_TRIP]: state => {
        const isRoundTrip = !state.isRoundTrip;
        return immutable(state, {
          isRoundTrip: { $set: isRoundTrip },
          flightDates: {
            return: { $set: isRoundTrip ? state.flightDates.return : null },
          },
          url: { $set: null },
        });
      },
      [ActionTypes.BOOKING_TOGGLE_PASSENGERS_DROPDOWN]: state => {
        const passengersDropdownIsOpen = !state.passengersDropdownIsOpen;
        return immutable(state, {
          passengersDropdownIsOpen: { $set: passengersDropdownIsOpen },
        });
      },
      [ActionTypes.BOOKING_CHANGE_PASSENGERS_NUMBER]: (state, { payload }) => {
        const total = Object.keys(payload.data).reduce((acc, key) => acc + payload.data[key], 0);
        return immutable(state, {
          passengers: { $set: Object.assign({}, state.passengers, payload.data, { total }) },
          url: { $set: null },
        });
      },
      [ActionTypes.BOOKING_URL_READY]: (state, { payload }) => {
        return immutable(state, {
          url: { $set: payload.url },
        });
      },
      [ActionTypes.BOOKING_SET_INVALID]: (state, { payload }) => {
        const invalidValues = ['from', 'to', 'departure'];
        return immutable(state, {
          invalid: { $set: invalidValues[payload.data] },
        });
      },
      [ActionTypes.BOOKING_UNSET_INVALID]: state => {
        return immutable(state, {
          invalid: { $set: null },
        });
      },
      [ActionTypes.BOOKING_CHANGE_FLIGHT_DATES]: (state, { payload }) => {
        return immutable(state, {
          flightDates: {
            departure: { $set: payload.departure },
            return: { $set: payload.return },
          },
          url: { $set: null },
        });
      },
    },
    bookingState,
  ),
};
