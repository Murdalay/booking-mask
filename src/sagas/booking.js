/**
 * @module Sagas/BookingMask
 * @desc BookingMask
 */

import { all, select, put, takeLatest } from 'redux-saga/effects';
import { ActionTypes } from 'constants/index';
import { constructUrlFromBooking } from 'modules/helpers';

/**
 * Get Repos
 */
export function* validateDepartureData() {
  try {
    const booking = yield select(state => state.booking);
    const valuesArray = [booking.from, booking.to, booking.flightDates.departure];
    const invalidIndex = valuesArray.findIndex(val => !val);

    if (invalidIndex > -1) {
      yield put({
        type: ActionTypes.BOOKING_SET_INVALID,
        payload: { data: invalidIndex },
      });
    } else {
      yield put({
        type: ActionTypes.BOOKING_SUBMIT_SUCCESS,
        payload: {},
      });
    }
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.EXCEPTION,
      payload: err,
    });
  }
}

/**
 * Get Repos
 */
export function* makeRedirect() {
  try {
    const booking = yield select(state => state.booking);
    const prefix = 'https://www.swiss.com/us/en/Book';
    const urlEnd = 'class-economy/al-LX/sidmbvl';
    const url = constructUrlFromBooking(booking, prefix, urlEnd);

    yield put({
      type: ActionTypes.BOOKING_URL_READY,
      payload: { url },
    });
  } catch (err) {
    /* istanbul ignore next */
    yield put({
      type: ActionTypes.EXCEPTION,
      payload: err,
    });
  }
}

/**
 * BookingMask Sagas
 */
export default function* root() {
  yield all([
    takeLatest(ActionTypes.BOOKING_SUBMIT, validateDepartureData),
    takeLatest(ActionTypes.BOOKING_SUBMIT_SUCCESS, makeRedirect),
  ]);
}
