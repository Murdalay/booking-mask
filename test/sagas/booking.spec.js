import { expectSaga } from 'redux-saga-test-plan';
import booking, { validateDepartureData, makeRedirect } from 'sagas/booking';
import { ActionTypes } from 'constants/index';

jest.mock('modules/client', () => ({
  request: () => ({ items: [] }),
}));

describe('booking', () => {
  it('should have the expected watchers', done =>
    expectSaga(booking)
      .run({ silenceTimeout: true })
      .then(saga => {
        expect(saga).toMatchSnapshot();
        done();
      }));

  it('validate departure saga should put BOOKING_SUBMIT_SUCCESS if state is valid', () =>
    expectSaga(validateDepartureData, { payload: {} })
      .provide({
        select() {
          const bookingValid = {
            from: 'FOO',
            to: 'BAR',
            flightDates: {
              departure: '11-11-2222',
            },
          };
          return bookingValid;
        },
      })
      .put({
        type: 'BOOKING_SUBMIT_SUCCESS',
        payload: {},
      })
      .run());

  it('validate departure saga should put BOOKING_SET_INVALID if `from` field has no value', () =>
    expectSaga(validateDepartureData, { payload: {} })
      .provide({
        select() {
          const bookingInalidFrom = {
            from: null,
            to: 'BAR',
            flightDates: {
              departure: '11-11-2222',
            },
          };
          return bookingInalidFrom;
        },
      })
      .put({
        type: 'BOOKING_SET_INVALID',
        payload: { data: 0 },
      })
      .run());

  it('validate departure saga should put BOOKING_SET_INVALID if `to` field has no value', () =>
    expectSaga(validateDepartureData, { payload: {} })
      .provide({
        select() {
          const bookingInalidTo = {
            from: 'FOO',
            to: null,
            flightDates: {
              departure: '11-11-2222',
            },
          };
          return bookingInalidTo;
        },
      })
      .put({
        type: 'BOOKING_SET_INVALID',
        payload: { data: 1 },
      })
      .run());

  it('validate departure saga should put BOOKING_SET_INVALID if `flightDates.departure` field has no value', () =>
    expectSaga(validateDepartureData, { payload: {} })
      .provide({
        select() {
          const bookingInalidDeparture = {
            from: 'FOO',
            to: 'BAR',
            flightDates: {
              departure: null,
            },
          };
          return bookingInalidDeparture;
        },
      })
      .put({
        type: 'BOOKING_SET_INVALID',
        payload: { data: 2 },
      })
      .run());

  it('should have the validate departure saga', () =>
    expectSaga(makeRedirect, { payload: { query: 'react' } })
      .provide({
        select() {
          const bookingValid = {
            from: 'FOO',
            to: 'BAR',
            flightDates: {
              departure: '11-11-2222',
            },
            passangers: {
              adults: 1,
              children: 0,
              infants: 0,
            },
          };
          return bookingValid;
        },
      })
      .put({
        type: ActionTypes.BOOKING_URL_READY,
        payload: {
          url:
            'https://www.swiss.com/us/en/Book/Outbound/FOO-BAR/from-11-11-2222/adults-1/children-0/infants-0/class-economy/al-LX/sidmbvl',
        },
      })
      .run());
});
