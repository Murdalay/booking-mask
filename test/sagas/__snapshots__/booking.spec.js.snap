// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`booking should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "BOOKING_SUBMIT",
          "@@redux-saga-test-plan/json/function/validateDepartureData",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "FORK": Object {
        "args": Array [
          "BOOKING_SUBMIT_SUCCESS",
          "@@redux-saga-test-plan/json/function/makeRedirect",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "BOOKING_SUBMIT",
      },
    },
    Object {
      "@@redux-saga/IO": true,
      "TAKE": Object {
        "pattern": "BOOKING_SUBMIT_SUCCESS",
      },
    },
  ],
}
`;
